grant-project/zcash-grant-system
# Anonymous contributions

Closes #247.

### What This Does

Makes it possible to create a contribution without a user, and fixes up a lot of aspects of contributions that relied on users. In detail:

* Enables the contribution form for logged out users
* Adds a checkbox to the contribution form that let's you contribute anonymously
* Adds a warning for anonymous contributions (of either type) that a user must agree to before contributing
* Adjusts some of the copy for contributions depending on whether or not you're anonymous
* Adds a stubbed out "Anonymous" user of userid 0 to contributions without users, as well as an `isAnonymous` boolean, _only on frontend_. The admin serializer doesn't add this spoofed user.
* Fixes up all backend code and emails to handle anonymous contributions
* Fixes up admin code to handle anonymous contributions
* (Unrelated) doesn't send contributions older than 24 hours to bootstrap, to keep it trim
* (Unrelated) downgrades blockchain's sentry library to improve stability

Some things I didn't do, but we might want to consider

* Delete anonymous contributions in the database after 24 hours. They can't be viewed by users, won't be checked by the blockchain checker on bootstrap, and are likely abandoned. Maybe up this to 2-3 days to be considerate, but we could clear these out for sure.
* Rate limit anonymous contributors. This would be kind of an annoying DDOS approach, so maybe we should limit anonymous contributors to something like 5 contributions per day per IP or something like that.
* Testing around anonymous contributions.
* Anonymous contributor icon. Zebra with a Groucho Marx glasses and nose, obviously.

### Steps to Test

* Open a live contribution when logged out and contribute, confirm the flow works as expected.
* Open a live contribution when logged in, contribute anonymously, confirm the flow works as expected.
* Open a live contribution when logged in, contribute normally, confirm the flow works as expected.
* View the proposal contributions tab, confirm the contributions render as expected.
* View your profile's funded tab, confirm you only have the attributed contributions.
* View the contributions tab in the admin, confirm the contributions show as expected.

## Screenshots

### Contribute anonymously checkbox (Only appears after typing in an amount)

<img width="316" alt="screen shot 2019-02-25 at 12 32 35 pm" src="https://user-images.githubusercontent.com/649992/53365958-7794a100-3910-11e9-86ee-ea959dbd5c9c.png">

### Anonymous contribution warning

<img width="560" alt="screen shot 2019-02-25 at 2 42 51 pm" src="https://user-images.githubusercontent.com/649992/53365979-7ebbaf00-3910-11e9-966d-c378f472e395.png">

### Contributions tab

<img width="656" alt="screen shot 2019-02-25 at 3 18 42 pm" src="https://user-images.githubusercontent.com/649992/53366073-c80bfe80-3910-11e9-8ec7-8f6b4aca735a.png">

### Admin

<img width="461" alt="screen shot 2019-02-25 at 3 18 17 pm" src="https://user-images.githubusercontent.com/649992/53366090-cf330c80-3910-11e9-9e44-78f58961ac8f.png">
